File: project_euler/01/014.go
// Copyright [2017] <dmnsn7@gmail.com>

package main

import "fmt"

const cn = 1000000

var dp = [cn]int{}

func f(n int64) int64 {
	if n%2 == 0 {
		return n / 2
	}
	return n*3 + 1
}

func dfs(n int64) int {
	if n == 1 {
		return 1
	} else if n < cn {
		if dp[n] == 0 {
			dp[n] = dfs(f(n)) + 1
		}
		return dp[n]
	}
	return dfs(f(n)) + 1
}

func main() {
	maxLen := 1
	target := 1
	for i := 1; i < cn; i++ {
		len := dfs(int64(i))
		if maxLen < len {
			maxLen = len
			target = i
		}
	}

	fmt.Printf("%d\n", target)
}

File: project_euler/01/014.py
#!/usr/bin/env python3
'''Copyright [2017] <dmnsn7@gmail.com>'''

N = 1000000

DP = [0 for _ in range(N)]


def dfs(num_n):
    if num_n == 1:
        return 1
    elif num_n < N:
        if DP[num_n] == 0:
            DP[num_n] = dfs(num_n // 2
                            if num_n % 2 == 0 else 3 * num_n + 1) + 1
        return DP[num_n]

    return dfs(num_n // 2 if num_n % 2 == 0 else 3 * num_n + 1) + 1


def main():
    max_length = 1
    target = 1
    for i in range(1, N):
        length = dfs(i)
        if max_length < length:
            max_length = length
            target = i

    print(target)


if __name__ == '__main__':
    main()
